File: code and data/MATLAB files generate Joint coordinate/Jerk.m
function [t, X, dX]=Jerk(nbre_pts, tf, P0, Pf)
  t=[0:tf/nbre_pts:tf];
  to=t/tf;
  for i=1:3
	X(i,:)=P0(i)+(P0(i)-Pf(i)).*(15*to.^4-6*to.^5-10*to.^3);
	dX(i,:)=(P0(i)-Pf(i)).*(4*15*to.^4./t-5*6*to.^5./t-3*10*to.^3./t);
    dX(i,1)=0; %dX(t=0) is undefined, should be 0
  end

  %figure();
  %plot3(X(1,:), X(2,:), X(3,:));
  %figure();
  %plot(t,dX);
end

File: scripts/python/isbulmodel/Jerk.py
# -*- coding: utf-8 -*-
"""
Created on Thu May  2 19:37:28 2024

Min jerk trajectory functions

@author: vcrocher - Unimelb
"""
import numpy as np

def Jerk(nbre_pts: int, tf: float, P0: np.array, Pf: np.array, plot=False):
    t=np.linspace(0, tf, nbre_pts)
    to=t/tf
    X = np.empty([3,nbre_pts])
    dX = np.empty([3,nbre_pts])
    ddX = np.empty([3,nbre_pts])
    for i, tto, tt in zip(range(0, nbre_pts), to, t):
        X[:,i]=P0+(P0-Pf)*(15*tto**4 -6*tto**5 -10*tto**3)
        dX[:,i]=(P0-Pf)*(4*15*tto**4/tt -5*6*tto**5/tt -3*10*tto**3/tt)
        ddX[:,i]=(P0-Pf)*(4*3*15*tto**4/(tt**2) -5*4*6*tto**5/(tt**2) -3*2*10*tto**3/(tt**2))
    dX[:,0]=[0,0,0]; #dX(t=0) is undefined, should be 0
    ddX[:,0]=[0,0,0]; #ddX(t=0) is undefined, should be 0

    if(plot):
        import matplotlib.pyplot as plt
        ax=plt.figure().add_subplot(projection='3d');
        ax.plot(X[1,:], X[2,:], X[3,:]);
        plt.figure()
        plot(t, dX)

    return t, X, dX, ddX
